include/lbann/utils/nvshmem.hpp

```diff
@@ -28,20 +28,88 @@
 #define LBANN_UTILS_NVSHMEM_HPP_INCLUDED
 
 #include "lbann/base.hpp"
+#include "lbann/utils/exception.hpp"
 #ifdef LBANN_HAS_NVSHMEM
-#include "nvshmem.h"
-#include "nvshmemx.h"
-
-#include "lbann/comm.hpp"
+#include <mpi.h>
+#include <nvshmem.h>
+#include <nvshmemx.h>
 
 namespace lbann {
 namespace nvshmem {
 
-/// Initialize NVSHMEM library
-void initialize(MPI_Comm comm);
-/// Finalize NVSHMEM library
+/** Whether NVSHMEM is active.
+ *
+ *  Returns true if NVSHMEM has been initialized and has not been
+ *  finalized.
+ */
+bool is_active() noexcept;
+
+/** @brief Initialize NVSHMEM library.
+ *
+ *  Does nothing if NVSHMEM has already been initialized. This is
+ *  _not_ thread-safe.
+ */
+void initialize(MPI_Comm comm=MPI_COMM_WORLD);
+
+/** @brief Finalize NVSHMEM library.
+ *
+ *  Does nothing if NVSHMEM has not been initialized or has already
+ *  been finalized. This is _not_ thread-safe.
+ */
 void finalize();
 
+/** @brief Allocate GPU buffer on the NVSHMEM symmetric heap.
+ *
+ *  Initializes NVSHMEM if needed.
```

Comment:
Hm, that's a good point. It will add some extra complexity for users though, since they'll need to remember to configure LBANN to initialize NVHSMEM rather than LBANN handling it automatically.

Reply:
Can this not be done in `lbann::initialize()` if `LBANN_HAS_NVSHMEM`?

Reply:
Do we want to initialize NVSHMEM every time though?

Reply:
That's a fair point. On second thought, let's leave this alone and have a discussion offline about cleaning up initialization. There are some other issues tied up in here, like the fact that runs that do use (NV)SHMEM will require more care since mallocs (that dispatch directly to `(nv)shmem_malloc`) are collective on the universe.